### ISSUE TITLE
join_with_return_assignment false positive on complex statements

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#1329]
Take the following code:

```dart
class A {
  int f;
  int g;
}

class B {
  final A a = A();
}

int main() {
  final B b1 = B();
  final B b2 = B();

  b1.a
    ..f = 1
    ..g = 1;
  b2.a.f = 2; // Error here
  return b1.a.f;
}
```

The `join_return_with_assignment` rule reports at the error comment "Join return statement with assignment when possible," because it does not see that `b1.a` is not equal to `b2.a`.

(Same cause as #1323)
